rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -0,0 +1,758 @@
+      param_server->parameter_list.data[i].name.data);
+  }
+}
+
+void
+rclc_parameter_server_get_service_callback(
+  const void * req,
+  void * res,
+  void * parameter_server)
+{
+  const GetParameters_Request * request = (const GetParameters_Request *) req;
+  GetParameters_Response * response = (GetParameters_Response *) res;
+  rclc_parameter_server_t * param_server = (rclc_parameter_server_t *) parameter_server;
+
+  size_t size = (request->names.size > param_server->parameter_list.size) ?
+    param_server->parameter_list.size :
+    request->names.size;
+
+  response->values.size = size;
+
+  for (size_t i = 0; i < size; i++) {
+    Parameter * parameter = rclc_parameter_search(
+      &param_server->parameter_list,
+      request->names.data[i].data);
+
+    if (parameter != NULL) {
+      rclc_parameter_value_copy(&response->values.data[i], &parameter->value);
+    } else {
+      response->values.data[i].type = RCLC_PARAMETER_NOT_SET;
+    }
+  }
+}
+
+void
+rclc_parameter_server_get_types_service_callback(
+  const void * req,
+  void * res,
+  void * parameter_server)
+{
+  const GetParameterTypes_Request * request = (const GetParameterTypes_Request *)  req;
+  GetParameterTypes_Response * response = (GetParameterTypes_Response *) res;
+  rclc_parameter_server_t * param_server = (rclc_parameter_server_t *) parameter_server;
+
+  if (request->names.size > response->types.capacity) {
+    response->types.size = 0;
+    return;
+  }
+
+  response->types.size = request->names.size;
+
+  for (size_t i = 0; i < response->types.size; i++) {
+    Parameter * parameter = rclc_parameter_search(
+      &param_server->parameter_list,
+      request->names.data[i].data);
+
+    if (parameter != NULL) {
+      response->types.data[i] = parameter->value.type;
+    } else {
+      response->types.data[i] = RCLC_PARAMETER_NOT_SET;
+    }
+  }
+}
+
+void
+rclc_parameter_server_set_service_callback(
+  const void * req,
+  void * res,
+  void * parameter_server)
+{
+  const SetParameters_Request * request = (const SetParameters_Request *) req;
+  SetParameters_Response * response = (SetParameters_Response *) res;
+  rclc_parameter_server_t * param_server = (rclc_parameter_server_t *) parameter_server;
+
+  if (request->parameters.size > response->results.capacity) {
+    response->results.size = 0;
+    return;
+  }
+
+  response->results.size = request->parameters.size;
+
+  for (size_t i = 0; i < response->results.size; i++) {
+    rosidl_runtime_c__String * message =
+      (rosidl_runtime_c__String *) &response->results.data[i].reason;
+    Parameter * parameter = rclc_parameter_search(
+      &param_server->parameter_list,
+      request->parameters.data[i].name.data);
+    rcl_ret_t ret = RCL_RET_OK;
+
+    if (parameter != NULL) {
+      response->results.data[i].successful = true;
+
+      if (parameter->value.type != request->parameters.data[i].value.type) {
+        rclc_parameter_set_string(message, "Type mismatch");
```

Comment:
This service info messages sizes may conflict with RCLC_PARAMETER_MAX_STRING_LENGHT value